### PR TITLE
Tenant qualify config store API

### PR DESCRIPTION
**Pull request by emswbandara**
Fixes https://github.com/wso2/product-is/issues/8545

### CODE REVIEW

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java [hunk 1]
```diff
@@ -652,9 +652,15 @@ public void addResource(Resource resource) throws ConfigurationManagementExcepti
...
-                                            .getTenantId());
+                            int tenantId;
+                            String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
+                            if (StringUtils.isBlank(tenantDomain)) {
+                                tenantId = PrivilegedCarbonContext.getThreadLocalCarbonContext()
+                                        .getTenantId();
+                            } else {
+                                tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+                            }
+                            preparedStatement.setInt(++initialParameterIndex, tenantId);
```
Comment (malithie) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java:663]: Is this needed. Isn't the expectation of the dao is to read tenant from PrivilegedCarbonContext.
In that case, we can have a tenant flow initiated from the service layer right

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -441,12 +482,25 @@ private void validateResourcesRetrieveRequest(String resourceTypeName)
 
     private int getTenantId() {
 
-        return PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
+        int tenantId;
+        String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
```
Comment (mefarazath) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java:486]: Can't we call getTenantDomain() method here?

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java [hunk 3]
```diff
@@ -987,8 +999,15 @@ private void updateMetadataForH2(Resource resource, String resourceTypeId, boole
...
-                                preparedStatement.setInt(initialParameterIndex,
-                                        PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId());
+                                int tenantId;
+                                String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
+                                if (StringUtils.isBlank(tenantDomain)) {
+                                    tenantId = PrivilegedCarbonContext.getThreadLocalCarbonContext()
+                                            .getTenantId();
+                                } else {
+                                    tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+                                }
```
Comment (mefarazath) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java:1009]: This logic seems to repeat. Can we move this logic to a common function / utility and re-use?

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java (hunk 1 above)
Comment (emswbandara) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java:663]: It seems that correct tenant domain is retrieved from PrivilegedCarbonContext even though we don't have a tenant flow initiated from the service layer. So shall I keep the old code as it is?

Comment (malithie) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java:663]: +1

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManager.java [hunk 4]
```diff
@@ -43,6 +43,17 @@
      */
     Resources getTenantResources(Condition searchCondition) throws ConfigurationManagementException;
```
Comment (tharindu-bandara) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManager.java:44]: shall we deprecate this?

Comment (emswbandara) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManager.java:47]: fixed

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.endpoint/src/main/java/org/wso2/carbon/identity/configuration/mgt/endpoint/impl/SearchApiServiceImpl.java [hunk 5]
```diff
@@ -42,14 +44,20 @@
...
         try {
-            Resources resources = getConfigurationManager().getTenantResources(
-                    getSearchCondition(searchContext)
-            );
+            Resources resources;
+            if (allowCrossTenantSearch) {
+                resources = getConfigurationManager().getTenantResources(getSearchCondition(searchContext));
+            } else {
+                resources = getConfigurationManager().getTenantResources(IdentityTenantUtil
+                                .getTenantDomainFromContext(), getSearchCondition(searchContext));
```
Comment (malithie) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.endpoint/src/main/java/org/wso2/carbon/identity/configuration/mgt/endpoint/impl/SearchApiServiceImpl.java:59]: If tenant qualified URL is not enabled will this have an impact?
IIRC, the thread-local gets populated on tenant qualified mode only, whereas this API is implemented to be accessible from a tenant in the path in the legacy way

Comment (emswbandara) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.endpoint/src/main/java/org/wso2/carbon/identity/configuration/mgt/endpoint/impl/SearchApiServiceImpl.java:59]: This worked as expected even when tenant qualified URL mode is disabled.

**Pull request merged**